archivessnake/create_daos.py

```diff
@@ -0,0 +1,205 @@
+parser.add_argument(
+    'dip_info_file',
+    help='Path to pickle file containing array of dictionaries with DIP information.')
+args = parser.parse_args()
+
+
+def main(resource, series, dip_file):
+    """Main function, which is run when this script is executed"""
+    aspace = ASpace(
+        baseurl=config.get(
+            "ArchivesSpace", "baseURL"), username=config.get(
+            "ArchivesSpace", "username"), password=config.get(
+                "ArchivesSpace", "password"))
+    dao_url = join(config.get("ArchivesSpace", "baseURL"),
+                   "repositories/2/digital_objects")
+    dip_data = load_dip_file(dip_file)
+    for uri in get_uris(aspace, resource, series):
+        try:
+            dip = match_dip_to_aip(aspace, uri, dip_data)
+            dip_uuid, title, aip_uuid = get_dip_info(dip)
+            dao_uri = post_dao(dao_url, aspace, dip_uuid, title, aip_uuid)
+            update_component(aspace, uri, dao_uri)
+        except Exception as e:
+            print(e)
+
+
+def load_dip_file(dip_file):
+    """Uses pickle to open external file of DIP information, stored as an array of dictionaries.
+
+    Args:
+        dip_info (str): path to pickle file
+
+    Returns:
+        DIP information (array)
+    """
+    dip_data = pickle.load(open(dip_file, "rb"))
+    return dip_data
+
+
+def get_uris(aspace, resource, series):
+    """Takes a resource and a series from that resource to return all file-level descendants of the series that do not have linked digital objects.
+
+    Args:
+        aspace (asnake.aspace.ASpace): instantiation of ASpace
+        resource (str): ArchivesSpace ID of resource. Example: 123
+        series (str): ArchivesSpace ID of series. Example: 4321
+
+    Returns:
+        list of ArchivesSpace URIs (array)
+    """
+    lsrm_tree = aspace.client.get(
+        "repositories/2/resources/{}/tree".format(resource)).json()
+    lsrm_tree.get('children')[1].get('children')
+    series_tree = [child for child in lsrm_tree.get(
+        'children') if child.get("id") == series][0]
+    list_of_lsrm_uris = []
+    for subseries in series_tree.get("children"):
+        for file_level in subseries.get("children"):
+            if "digital_object" not in file_level.get("instance_types"):
+                list_of_lsrm_uris.append(file_level.get('record_uri'))
+    return list_of_lsrm_uris
+
+
+def match_dip_to_aip(aspace, ao_uri, dip_list):
+    """Using box and folder indicators, matches an ArchivesSpace component to a DIP.
+
+    Args:
+        aspace (asnake.aspace.ASpace): instantiation of ASpace
+        ao_uri (str): ArchivesSpace URI of an archival object
+        dip_list (array): List of dictionaries of DIP information
+
+    Returns:
+        DIP information (dictionary)
+    """
+    ao = aspace.client.get(
+        ao_uri, params={"resolve": ["top_container"]}).json()
+    box = ao.get('instances')[0].get('sub_container').get(
```

Comment:
A note on `get()` - the benefits of this are if you aren't 100% sure what the structure of your data looks like and/or if that structure is variable. In a couple of cases here, the data will ALWAYS have a particular key, so you can just reference a key directly.